src/abacus-ts/websocket/lib/indexerWebsocket.ts

```diff
@@ -117,80 +176,144 @@ export class IndexerWebsocket {
-        'unsubbing from nonexistent or already unsubbed channel',
-        { channel, id, wsId: this.indexerWsId }
-      );
+
+    if (forceSuppressWsMessage) {
       return;
     }
+
     logAbacusTsInfo('IndexerWebsocket', 'removing subscription', {
       channel,
       id,
       socketNonNull: this.socket != null,
       socketActive: this.socket?.isActive(),
       wsId: this.indexerWsId,
     });
-    if (
-      this.socket != null &&
-      this.socket.isActive() &&
-      this.subscriptions[channel][id ?? NO_ID_SPECIAL_STRING_ID]!.sentSubMessage
-    ) {
+
+    if (this.socket != null && this.socket.isActive() && sub.sentSubMessage) {
       this.socket.send({
         channel,
         id,
         type: 'unsubscribe',
       });
     }
-    delete this.subscriptions[channel][id ?? NO_ID_SPECIAL_STRING_ID];
   };
 
   private _refreshSub = (channel: string, id: string | undefined) => {
-    const sub = this.subscriptions[channel]?.[id ?? NO_ID_SPECIAL_STRING_ID];
+    const sub = this.subscriptions.getSubscription(channel, id);
     if (sub == null) {
       return;
     }
     this._performUnsub(sub);
     this._addSub(sub);
   };
 
-  // if we get a "could not fetch data" error, we retry once as long as this channel+id is not on cooldown
-  // TODO: remove this entirely when backend is more reliable
-  private _handleErrorReceived = (message: IndexerWebsocketErrorMessage) => {
+  private _maybeRetryTimeoutError = (message: IndexerWebsocketErrorMessage): boolean => {
     if (message.message.startsWith('Internal error, could not fetch data for subscription: ')) {
       const maybeChannel = message.channel;
       const maybeId = message.id;
-      if (maybeChannel != null && maybeChannel.startsWith('v4_')) {
-        const channelAndId = `${maybeChannel}${CHANNEL_ID_SAFE_DIVIDER}${maybeId ?? NO_ID_SPECIAL_STRING_ID}`;
-        const lastRefresh = this.lastRetryTimeMsByChannelAndId[channelAndId] ?? 0;
-        if (Date.now() - lastRefresh > CHANNEL_RETRY_COOLDOWN_MS) {
-          this.lastRetryTimeMsByChannelAndId[channelAndId] = Date.now();
+      if (
+        maybeChannel != null &&
+        maybeChannel.startsWith('v4_') &&
+        this.subscriptions.hasSubscription(maybeChannel, maybeId)
+      ) {
+        const sub = this.subscriptions.getSubscription(maybeChannel, maybeId)!;
+        const lastRefresh = sub.lastRetryBecauseErrorMs ?? 0;
+        const hasBaseData = sub.receivedBaseData;
+        if (!hasBaseData && Date.now() - lastRefresh > CHANNEL_RETRY_COOLDOWN_MS) {
+          sub.lastRetryBecauseErrorMs = Date.now();
           this._refreshSub(maybeChannel, maybeId);
-          logAbacusTsInfo('IndexerWebsocket', 'error fetching data for channel, refetching', {
+          logAbacusTsInfo('IndexerWebsocket', 'error fetching subscription, refetching', {
             maybeChannel,
             maybeId,
+            socketNonNull: this.socket != null,
+            socketActive: this.socket?.isActive(),
```

Comment:
should this be a boolean return as well for clarity in DD? `Boolean(this.socket?.isActive())`